clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -258,7 +258,10 @@ public void poll(long now) {
         }
 
         pollHeartbeat(now);
-        maybeAutoCommitOffsetsAsync(now);
+        if (subscriptions.hasNoSubscription())
+            subscriptions.clearAssignment();
```

Comment:
Yes, I was debating that myself. Then I thought if a consumer is subscribed to topic `foo`, the behavior should be the same in the following scenarios that would happen next:

- calling `subscribe(Collections.singleton("bar"))`
- calling `unsubscribe()` first, and then `subscribe(Collections.singleton("bar"))`

If we clear the assignment right away the above use cases would lead to slightly different behavior. Unless that's Okay.